### PR TITLE
Allow vecX uniforms and inputs

### PR DESCRIPTION
**Pull request by oscons**
Hi @ojack ,

some transforms take a `vec4` as an input and while one could potentially use a `solid` as an input that would only work for `combine` and `combineCoord` type of functions.

The proposed changes would allow for generic `vec` arguments that could either be uniforms or static vectors. This would make specifying multiple vectors as arguments (e.g. colors) and binding to uniforms less verbose IMHO. Here's an example:

```javascript
// transform definition
const myTransforms = {
  linGrad: {
    type: 'src',
    inputs: [
      {name:'cstart',type:'vec4',default:'vec4(1.0,0.0,0.0,1.0)'},
      {name:'cend',type:'vec4',default:'vec4(1.0,0.0,0.0,1.0)'}
    ],
    glsl: `vec4 linGrad(vec2 _st, vec4 cstart, vec4 cend){
  return cstart + (cend - cstart) * sqrt(_st.x * _st.x))
}`
}}

// hydra call

const ft = ({time}) => [0, time % 2]

shape(3)
  .modulateScale(
    linGrad([0,1.0], ft)
      .repeat(10,10)
      .rotate(10)
  ).out(o0)
```

Let me know if you're interested in merging this and if anything needs to be changed.

Thanks. @oscons

### CODE REVIEW

**Comment by ojack** [comment]
I haven't tried yet, but looks great to me!

Another thing along these lines that I have wanted to try is making it possible for ANYTHING to use a texture as an input. So basically you could do something like:
´
osc(10, 0, src(s0)).out()
´

And then the last parameter would be affected by the colors in the input source. It would make the modulate functions unnecessary because then everything would be modulate-able. The main question I have about this from a usability perspective is how to flexibly map the r,g,b parameters to the input parameters of the function in a way that isn't too confusing.

**Comment by oscons** [comment]
Thanks for merging!

Your comment regarding the texture input for anything is something I'm currently looking into (and one of the reasons for #23 btw). My current approach is to create a new `modX` function for all `coord` and `src` type transforms (for now) and replace all `float` arguments with a `vec4` that describes the scale (in `rg`) and offset (in `ba`) for the modulation by the `xy` of the input texture.

```c
vec4 modShape(vec2 st, vec4 c0, vec4 sides...) {
  float m_sides = (c0.r * sides.r + sides.b) + (c0.g * sides.g + sides.a);
  return shape(st, m_sides, ...)
}
```

Together with the functionality from this pull request here (and one more I might open, regarding a new transformation type), it allows you to write something like this (option _A_):

```javascript
s0.init({src: someImageOrVideo})
modShape(src(s0), [2, -0.5, 2, 0]).out(o0)
```

The one thing about this approach I dislike the most though is that you'd loose the ability to do modulation based on the `b` and `a` component of the incoming texture/vec4. So I'm also experimenting with generating functions and/or parameters to make the glsl and Hydra expressions less related in their definition. The downside I currently see in that is that i might become quite verbose. Here's one approach I had in mind (option _B_):

```javascript
shape({mod: src(s0), sides: "x*2+2+y*-0.5+0"})
// alternatively
shape().mod(src(s0), {sides: "x*2+2+y*-0.5+0"})
```

This would translate to the following generated glsl:
```c
vec4 shape(vec2 st, vec4 mod_sides, ...) { // one function needs to be generated per expression definition
  float m_sides = mod_sides.x * 2.0 + 2.0 + mod_sides.y * (-0.5) + 0; // generated from the parsed expression
  return shape(st, m_sides, ...);
}

```

Another approach would be similar to the first one with the `modShape` but making the glsl fully parameterized and then somehow come up with a short notation/helper function for the Hydra part, e.g. the following glsl (option _C_):
```c
vec4 shape(vec2 st,
  vec4 mod_sides, vec4 mod_sides_parm_r, vec4 mod_sides_parm_g, vec4 mod_sides_parm_b, vec4 mod_sides_parm_a, ...) {
  float sides = mod_sides.r * mod_sides_parm_r.x // scale
         + mod_sides_parm_r.y  // post-scale-offset
         +  mod_sides.g * mod_sides_parm_g.x + mod_sides_parm_g.y
         +  mod_sides.b * mod_sides_parm_b.x + mod_sides_parm_b.y
         +  mod_sides.a * mod_sides_parm_a.x + mod_sides_parm_a.y;
   // mod_sides_parm_?.za not used, could be reduced to vec2 but za might come in usesfurl for something else
  // also, there might be some vector math shortcut here, haven't thought hard enough about it
  return shape(st, m_sides, ...);
}
```
And the Hydra part:
```javascript
shape(mod(src(s0), [2, 2], [-0.5, 0], [0, 0], [0, 0])) // modulator and RGBA[scale, offset] parameters
shape(mod(s0, [2, 2], [-0.5])) // same, but with defaults used
// could be combined with "static" parameters
shape(mod(s0, [2, 2], [-0.5]), 0.5, mod(voronoi().hue(), 0.1)) // mod parameters can be "smart" about their type, here the '0.1' would mean "scale 0.1, offset default (=0)"
```

Basically `shape` and other functions in Hydra need to be "overloaded" to handle the complexity. 

I currently experiment with option A because it's easy to implement, however I'll soon give option C a try as well.

Have you already looked into implementing something like this? Do you have any suggestions/preferences? Some things I might not have considered?

**Comment by oscons** [comment]
Fyi, here's a [link to an implementation of option A](https://oscons.github.io/hydra-synth/#eNqdVm2P2zYM/ivcIailnqvY2Q4YckiHoSj2scCKAQPiLPbZcuLCkQ1Z3lqk/u8jJb8ll+uHcxDF4juph1TOd/JufVfk7Ke9bhUHLU2rFTT4UzP+GKlIHXSShbABxmHzHnDzB35ZpAC2oQ/BDpZLaEyiDZF82L4LxIMPK0uXKosUF02qq7L8mwUifOBCVyYxkgUoNLHeWd7oj3FRtYZVgSWdquzzMaml9drze29h4P/qQigy2fRENE3mcPUDETxYPmoV7SQQiJXjO+VTVZljoQ4UrXUcDo4/Nekttz5QxJgClWAi0lY87EYzP1sz9NFYC6mZ3S9f9bxeL1J5q1JTVOPJwpli1hK3++O3TCeM5aqxJ2w5ALSmlWoMKPlfrhAA545o9P309EWmRkhldCEbq+mURF6UBrNkWx9ytbP2kC3Mt1rCZrMBL60qnXnw/TtckBudevDmDRELVbemEaVUB3OE9xDw0XilPybpEa2ryXwf7hArBRrjsS3OSjTtU2M0Ai3kwlR/1bXUH5JGMt7NuCHv4ksTBZqY4jglNbOvc2f4FDk4+iyLvKwS4/G5GD19U11RHQfN1zJbg9Gt9J8LYCZ7lZzkGpPaL87OIRG6+Ia0Za9vOaLH2ZlM+LfFKJ01eP/K9BfvBZFM5klboqeYpLCtbcPhMkTYC2DVPyNE1AHbWcu6TFLJlv+wKMru+WLpe4tQeLyzujx+7qq7JM22Hciyka+pc56g4g8L/cMCWebLYbnXjk8tNKAqV3uHkxu4/43a4vRUKPl7WXqwHrcfbLNcmcEsX7Jizwzc4a2u9fLia6IPqGrFIMVpdS2SJmW574U8GhuzrsjxTGkITNk6oPSJzWrlGmcN24uDcMXFzMpKX+JqDrhZLXcCPaeJYUIIVdg+ZA5enW18+8r5zNShbErEZKQWZywP7UbU4QyDP+Xh49eaxQw9RfhktDT3nOE06Oz7W1oZjzmWZhESJNL9YoUQpVqQUSp+hyhXqiMrK9g3hlA/lpcmy41Y46ExKNcOLhuZiy9VoZiHTnlnh4c1MozS8wBeZ2rYDUNx5ssfUdx7tfMIvY1kPMAL30LDW0QC/tzD0ID3U7RO5glJl5SD0zo8YySPQzZxFKmYCtc3ZV9MrDSbajbgbV61qxRmwV8VCv+cjJPbdt7wbu+rrVI7zNbC9nHelu7eszFdXWNWry8rIHKyFoHDjnixK/Cvrxs4bq2DXA2E3uixd0Y3FF2L5NXi5677H67du7g=).

**Comment by ojack** [comment]
How do you run code in your sandbox environment? :)

There are things I like about all of these ideas. I want to play around with them more. Some thoughts:

1) I think the ability scale and offset the rgba values, as well as swap channels are functions that are useful in hydra in general. Actually maybe they already exist with add() and mult() functions in hydra,   and the vec4 functionality that you have just added. To what extent are the following (from your option A) describing the same thing?
```
modShape(src(s0), [2, -0.5, 2, 0]).out(o0)
```
```
modShape(src(s0).add([2, 0]).mult([2, -0.5])).out(o0)
```
There is something that I like about the second as it sticks with existing hydra functions and also lets you visualize the scaling and offset of the modulation as colors. I think there are definitely functions missing that would make this usable. For example: 
```
src(s0).b().out()
```
could be a possible function that replaces the value of all of the channels with the b channel. Then it would be easy to specify which channel to use for modulation. 

2) I really like the idea in option b of being able to pass in equations as a string, and I wonder to what extent that could be applicable in general. Gonna think about it more :)

**Comment by oscons** [comment]
`Shift-Ctrl-Enter` should execute the code, provided you have the focus on the editor input. If nothing's happening, maybe I've managed to build in something that breaks in your browser, though recent Chrome and FF should work. Just let me know and I'll fix it.

Regarding the two `modShape` statements, there's a subtlety in how I implemented modShape that makes the second statement not work:

```c
float sides = mod_sides.a * mod_sides_parm_r.x + mod_sides_parm_r.y ... // plus the same for gba components
```

The default for `mod_sides_parm_*` is `[0, 0, 0, 0]` which means that the texture will not influence the parameter. I've chosen this default, since i thought it's more likely that you don't want to modulate all parameters, an hence a default of `[0, 0, 0, 0]` means less typing. On the other hand this also means you'd need to add the initialization values explicitly to your second statement:

```javascript
modShape(src(s0).mult([2, -0.5]).add([2, 0]), [1, 1]).out(o0)
```

Now, that's not too far off something like this

```javascript
hydra.synth.setFunction('sum', {
      type: 'color',
      inputs: [{name: 'mult', type: 'vec4', default: [1, 1, 1, 1]}],
      glsl: `float sum(vec4 c0, vec4 _mult){
  vec4 v = c0 * _mult;
  return v.r + v.g + v.b + v.a;
}`})

shape(src(s0).mult([2, -0.5, 0 ,0]).add([2, 0]).sum()).out(o0) // the ba component on mult is currently necessary, since the default is not set and the vecX default is [0, 0, 0, 1]
````

And that's basically another way of doing what I had in mind for option C, just with the `sum` instead of the `mod`. Actually I think `sum` might be a sane default for converting `vec4` to `float`, so it could be implicitly added, which reduces the typing required.

That `sum` implementation above will not work on the current `dev` branch at the moment but I'll open another pull soon for a feature that should make this work as well.
